### PR TITLE
VPN-3512 - [PROPOSAL] Require C++ 20 for builds

### PR DESCRIPTION
**Pull request by brizental**
The motivation for this change is that I would like to have support for [aggregate initializion](https://en.cppreference.com/w/cpp/language/aggregate_initialization) to have the nice API support for Glean event extras (see: #5148). In practice I would like to be able to initialize structs like:

```cpp
Struct {
   .a = "a",
   .b = "b"
}
```

Instead of:

```cpp
Struct struct;
struct.a = "a";
struct.b = "b";
```

Ok, that may not sound like strong reasoning to make such a big change. However, is it really that big of a change for us?

1. From some initial tests it seems like this "just works". I could not find much on the web about Qt and C++20, but we are not building it from scratch anyways so does it matter?
2. All of the platforms are building fine, there is some issue with the Linux CI but the `linux/opt` build did not fail so that is probably something with the specific jobs and I am trying to address it.
3. The API is way better for the Glean event recording with the aggregate initializers and it is the exact same as the Firefox Desktop API so I would love to keep it. I am strongly attempting to keep the APIs exactly the same so we can 1. use their documentation 2. not diverge too much from their Glean implementation, the more we diverge the less we can share the implementation.
4. Using a newer version of the language just sounds likea good call overall?

Ok, I'm done. WDY'allT?

Note:  see [Annex C](https://open-std.org/JTC1/SC22/WG21/docs/papers/2020/n4861.pdf) for compatibility notes on C++20.

### CODE REVIEW

**Comment by oskirby** [comment]
I dunno, I have mixed feelings on making this change. To my knowledge, we can turn this on relatively easily for every platform except Ubuntu 18.04/Bionic. And as this implemented, this is actually forcing a downgrade of the C++ compiler on Ubuntu 20.04/Focal and beyond.

We have had to workaround a few stale tool versions on Bionic before (eg: golang and cmake), but we handled them in a slightly more precise way, by adding these two tweaks:
1. In the `linux/debian/control` file, we add a GCC version dependency like `gcc (>= 8.0) | gcc-8` which will leave the GCC compiler as default if it's greater than version 8, or we request a specific version of GCC.
2. In the `linux/debian/rules` file, we add a check if a sufficient version of gcc exists, and set the `CC` and `CXX` variables accordingly if not.

That being said, I would prefer that we just wait until April 2023 when we can just drop Ubuntu/Bionic altogether.

**Comment by brizental** [comment]
>That being said, I would prefer that we just wait until April 2023 when we can just drop Ubuntu/Bionic altogether.
 
@oskirby Is there a concern that this will break the build? Or is the issue more on adding special cases for a platform we will not support in just a few months?

**Comment by oskirby** [comment]
I am concerned that the compiler downgrade is going to negatively impact our builds on newer Ubuntu releases. And I would like to request changes so that we don't force a downgrade unnecessarily.

I can be convinced that we should put in the effort to workaround this for Ubuntu bionic if you think it's necessary for the glean work. But you are right that we will not need the workaround in 4 months time.

**Comment by brizental** [comment]
> I am concerned that the compiler downgrade is going to negatively impact our builds on newer Ubuntu releases. And I would like to request changes so that we don't force a downgrade unnecessarily.

Ok, with regards to that... The truth I did not even realize I was doing it :/ I will attempt to do what you suggested in the previous comment.

> I can be convinced that we should put in the effort to workaround this for Ubuntu bionic if you think it's necessary for the glean work. But you are right that we will not need the workaround in 4 months time.

I think the pros of doing the upgrade are greater than the con of implementing the Bionic workaround. I say this without having succesfully implemented the workaround yet, but from your descritpion it seems straight forward? Let me know if I am being too optimistic here.

**Comment by oskirby** [comment]
@brizental Yeah, let's give the workaround a try by messing around with the `control` and `rules` files. If it just works, then let's merge it.

**Review by strseb — APPROVED** [review]
r+

r+ - We were going to remove the networkwatchers anyway - so cutting winrt dep's makes sense :)

File: src/platforms/windows/windowsnetworkwatcher.cpp [hunk 1]
```diff
@@ -148,26 +139,6 @@ void WindowsNetworkWatcher::processWlan(PWLAN_NOTIFICATION_DATA data) {
...
-    return TransportType_Unknown;
-  }
-
-  switch (device.IanaInterfaceType()) {
-    case 6:
-      return TransportType_Ethernet;
-    case 71:
-      return TransportType_WiFi;
-  }
+  // TODO: Implement this once we update to Qt6.3 (VPN-3511)
```
Comment (bakulf) [inline src/platforms/windows/windowsnetworkwatcher.cpp:142]: why this change?

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/platforms/windows/windowsnetworkwatcher.cpp (hunk 1 above)
Comment (brizental) [reply src/platforms/windows/windowsnetworkwatcher.cpp:142]: I am not sure you saw the message I left on the commit explaining the change, so from the commit message:

```
The version of cppwinrt that we use is v2.0.200609.3
and it comes from Windows SDK 10.0.19041.0.

That version of cppwinrt has a bug for suppoting C++20,
which is fixed on version 2.0.200921.6.

It's unclear wheather it is safe for us to update to a Windows SDK
version that includes the correct version of cppwinrt, because that
would require us to be on a Windows 11 SDK. We are unsure whether
building with that will result in binaries that can also be run on
Windows 10.

After speaking to Basti, we decided the least problematic way to address
this is to simply remove the code that relies on cppwinrt. That will be
removed anyways as soon as we more to Qt6.3 or above, because Qt will
provide the APIs we are currently getting from cppwinrt.

That code is only being called to add an extra key to a Glean event
recording, so this doesn't actually affect user facing features on the
app.
```

Let me know if that is still unclear.

**Review by birdsarah — APPROVED** [review]
I don't have enough knowledge of the build stuff this PR touches to provide detailed feedback but I'm supportive of the change and appreciate the documentation..

File: linux/debian/rules [hunk 2]
```diff
@@ -11,6 +11,12 @@ ifneq (ok,$(dpkg --compare-versions $(GOLANG_NATIVE_VERSION) ge 2:1.13 && echo o
 	export PATH := $(GODIR)/bin:$(PATH)
 endif
 
+GCC_NATIVE_VERSION = $(shell gcc -dumpversion 2>/dev/null || echo 0)
+ifneq (ok,$(shell dpkg --compare-versions $(GCC_NATIVE_VERSION) ge 8 && echo ok))
```
[the remark this answers is not in the document]
Comment (oskirby) [reply linux/debian/rules:15]: That's a good catch! I am indifferent on what path to take regarding the golang hack, since it's only needed for Ubuntu/Bionic and we're going to be deprecating it in April anyways. Please feel free to either fix it in this PR or just ignore the problem it goes away in April.

[the remark this answers is not in the document]
Comment (brizental) [reply linux/debian/rules:15]: I'll... just ignore it then :eyes:

[the remark this answers is not in the document]
Comment (brizental) [reply linux/debian/rules:15]: Hm, actually this is downgrading go for all verisons of Ubuntu, no? We always enter the conditional. I'll file a ticket and a separate PR and we can discuss it there.
